Feature/storage/transactional crc

jaschrep-msft opened the pull request:
- Introduces custom Storage CRC64 calculator, prepared to inherit from the future-released [NonCryptographicHashAlgorithm](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Hashing/src/System/IO/Hashing/NonCryptographicHashAlgorithm.cs)
- Converts every upload/download method in Blobs, Files.Datalake, and Files.Shares to
  - use options bags if they weren't already
  - use one of the new `UploadTransactionalHashingOptions` or `DownloadTransactionalHashingOptions`
- Hashing is opt-in but defaults to CRC after that opt-in
- After opt-in, SDK does calculations for you. This is opt-out (cannot opt out on more complex methods)

jaschrep-msft committed:
Transactional hashing models

jaschrep-msft committed:
StageBlock now has transactional hashing support

jaschrep-msft committed:
ContentHasher supports precalculated hash

GetHash will now read all options and pull a possible precalculated hash
into its response instead of calculating the result from scratch

jaschrep-msft committed:
New overloads

appendblobclient and blockblobclient now have options bag overloads for
their baseline upload methods.

jaschrep-msft committed:
Append and page blob updates

Those clients now have new APIs for base uploads, deprecated old
overloads, and updated writestreams to use the new internal signature.

jaschrep-msft committed:
Alligned XxxOpenWriteOptions for hashing options

jaschrep-msft committed:
Blob Write Streams Updated

jaschrep-msft committed:
uploadTransactionalHashing in datalake and shares

jaschrep-msft committed:
fixed mock issues

Fixed how some mocks were constructed due to new signatures for some
internal methods

jaschrep-msft committed:
blob download range transactional hashing

jaschrep-msft committed:
crc46 implementation and download tests

jaschrep-msft committed:
several read methods updated to transactional hash

jaschrep-msft committed:
testing

jaschrep-msft committed:
datalake/share download

jaschrep-msft committed:
build fix

jaschrep-msft committed:
renames | Blob/datalake downloadTo hashing

jaschrep-msft committed:
tests and fixes

jaschrep-msft committed:
Range required for download hashing

testing, recording.

jaschrep-msft committed:
more tests. property renames for consistency.

jaschrep-msft committed:
Azure.Storage.Models => Azure.Storage

Moved transactional hashing options classes down a namespace

jaschrep-msft committed:
fixes and rerecording transactional hashing tests

jaschrep-msft committed:
recording fix

jaschrep-msft committed:
recording

jaschrep-msft committed:
test fixes

jaschrep-msft committed:
all hashing tests scan messages for headers

jaschrep-msft committed:
minor fixes and test organization

jaschrep-msft committed:
test fine-tuning and rerecording

jaschrep-msft committed:
open write tests refactored for reusability

jaschrep-msft committed:
test refactors

jaschrep-msft committed:
More test refactoring

jaschrep-msft committed:
refactored parallel upload tests

jaschrep-msft committed:
finished blob test refactoring

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into feature/storage/transactional-crc

jaschrep-msft committed:
exportapi

jaschrep-msft committed:
fix export

jaschrep-msft committed:
fixed a scope issue

jaschrep-msft committed:
fixed datalake.append scope and common mock

jaschrep-msft committed:
fixed mock setup

jaschrep-msft committed:
More mocking issues

jaschrep-msft committed:
irerecord

jaschrep-msft committed:
full rerecording transactional hash tests

jaschrep-msft committed:
revert force record mode

jaschrep-msft committed:
testing CI failure

removing an inserted policy from hashing partitioned download tests to
determine  if it is interfering in recording interpretation.

jaschrep-msft committed:
Revert "testing CI failure"

This reverts commit 4bbb8d378b548a6e59d0b416d5baa6b3f8124146.

jaschrep-msft committed:
made encryption/hashing test liveonly

jaschrep-msft committed:
datalake transactional hashing tests

jaschrep-msft committed:
port hashing tests to file shares

jaschrep-msft committed:
transactional hash arg validation

jaschrep-msft committed:
parallel upload doesn't accept precalculated hash

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into feature/storage/transactional-crc

jaschrep-msft committed:
rerecord transactional hashing tests

jaschrep-msft committed:
restored accidental recording delete

jaschrep-msft committed:
rerecorded OpenRead tests

jaschrep-msft committed:
rerecord misc file share tests

jaschrep-msft committed:
New storagecrc64 API shape

storage crc calculator prepared to inherit NonCryptographicHashAlgorithm
when released.
Static crc calculator converted to use ReadOnlySpan<byte> instead of
byte[].
Static crc calculator converted to use BinaryPrimitives instead of
BitConverter. NOTE this forces the calculation into a specific
endianness which is acceptable based on CRC properties.

jaschrep-msft committed:
exportapi

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -315,6 +323,23 @@ public partial class BlobAppendInfo
         public bool IsServerEncrypted { get { throw null; } }
         public System.DateTimeOffset LastModified { get { throw null; } }
     }
+    public partial class BlobBaseDownloadOptions
```
Arch board might want us to name this `BlobDownloadOptions`, I don't think we currently have any model classes with `Base` in the name.

jaschrep-msft replied:
This name was present in an APIView they looked at.

@KrzysztofCwalina do you have a preference on this? The client name is `BlobBaseClient`, so this name does match current patterns, but I can understand an exception for this case.

kasobol-msft replied:
+1 for `BlobDownloadOptions`

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -315,6 +323,23 @@ public partial class BlobAppendInfo
+        public Azure.HttpRange Range { get { throw null; } set { } }
+        public Azure.Storage.DownloadTransactionalHashingOptions TransactionalHashingOptions { get { throw null; } set { } }
+    }
+    public partial class BlobBaseDownloadToOptions
```
Same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -1079,6 +1105,7 @@ public partial class BlobUploadOptions
         public System.Collections.Generic.IDictionary<string, string> Metadata { get { throw null; } set { } }
         public System.IProgress<long> ProgressHandler { get { throw null; } set { } }
         public System.Collections.Generic.IDictionary<string, string> Tags { get { throw null; } set { } }
+        public Azure.Storage.UploadTransactionalHashingOptions TransactionalHashingOptions { get { throw null; } set { } }
```
Why is the class name `UploadTransactionalHashingOptions` instead of `TransactionalHashingOptions`?  In the context of `BlobProperties` this seems a little odd to me.

jaschrep-msft replied:
There is also a `DownloadTransactionalHashingOptions` that goes on download methods and has some slightly different properties. So the class names need to be differentiated.

I also don't understand what you mean by "in the context of `BlobProperties`."

seanmcc-msft replied:
As a users, I call `BlobClient.GetProperties()`, which has a `UploadTransactionalHashingOptions` on it.  This seem like the name could be confusing, they might be wondering why stuff associated with uploading has been permanently stored in Blob Properties.

kasobol-msft replied:
@seanmcc-msft I believe the confusion happens because of github collapsing sections of the diff. This isn't nested under `BlobProperties` as far as I can tell.

Maybe unstead of `Upload/DownloadTransactionalHashingOptions` we should have something that mentions where the validation occurs. I.e. `ServerSideTransactionalHashingValidationOptions` or something along these lines.

seanmcc-msft replied:
@kasobol-msft, that is correct, I was confused by the UI.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -1241,6 +1276,7 @@ public partial class PageBlobOpenWriteOptions
         public Azure.Storage.Blobs.Models.PageBlobRequestConditions OpenConditions { get { throw null; } set { } }
         public System.IProgress<long> ProgressHandler { get { throw null; } set { } }
         public long? Size { get { throw null; } set { } }
+        public Azure.Storage.UploadTransactionalHashingOptions TransactionalHashingOptions { get { throw null; } set { } }
```
Same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1118,35 +1207,38 @@ public new AppendBlobClient WithEncryptionScope(string encryptionScope)
         {
             using (ClientConfiguration.Pipeline.BeginLoggingScope(nameof(AppendBlobClient)))
             {
+                options ??= new AppendBlobAppendBlockOptions();
```
Nit: we can remove this line and reference `options` as `options?` below.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1118,35 +1207,38 @@ public new AppendBlobClient WithEncryptionScope(string encryptionScope)
 
                 // All AppendBlobRequestConditions are valid.
-                conditions.ValidateConditionsNotPresent(
+                options.Conditions.ValidateConditionsNotPresent(
```
`ValidateConditionsNotPresent()` works on null conditions, IIRC.  Is this correct?

jaschrep-msft replied:
Correct. The extension method does a null check on the `this` parameter.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Azure.Storage.Blobs.csproj:
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Is there any change on this line?

jaschrep-msft replied:
I do not see a difference. This may be diff weirdness.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1091,43 +1108,148 @@ private void SetNameFieldsIfNull()
+        ///     </item>
+        /// </list>
+        /// </remarks>
+        [ForwardsClientCalls]
```
We might be able to remove this annotation if we create a new DiagnosticScope in this method.

jaschrep-msft replied:
It looks as if we can just remove these regardless. They must have been leftover from some struggles I was having.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1091,43 +1108,148 @@ private void SetNameFieldsIfNull()
+        ///     </item>
+        /// </list>
+        /// </remarks>
+        [ForwardsClientCalls]
```
Same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1559,25 +1691,126 @@ private void SetNameFieldsIfNull()
+        ///     </item>
+        /// </list>
+        /// </remarks>
+        [ForwardsClientCalls]
```
Same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1559,25 +1691,126 @@ private void SetNameFieldsIfNull()
+        ///     </item>
+        /// </list>
+        /// </remarks>
+        [ForwardsClientCalls]
```
Same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/AppendBlobAppendBlockOptions.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Azure.Storage.Blobs.Models
+{
+    /// <summary>
+    /// Optional paratmers for appending a block.
```
Typo, also we can add a reference here to the `AppendBlobClient.AppendBlockAsync()` which takes this overload.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/BlobBaseDownloadOptions.cs:
```diff
@@ -0,0 +1,37 @@
+namespace Azure.Storage.Blobs.Models
+{
+    /// <summary>
+    /// Optional paratmers for downloading some range of a blob.
```
Typo.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/BlockBlobStageBlockOptions.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Azure.Storage.Blobs.Models
+{
+    /// <summary>
+    /// Optional paratmers for staging a block for a block blob.
```
Typo.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/PageBlobUploadPagesOptions.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Azure.Storage.Blobs.Models
+{
+    /// <summary>
+    /// Optional paratmers for uploading pages.
```
Typo.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/tests/Azure.Storage.Blobs.Tests.csproj:
```diff
@@ -30,4 +30,7 @@
     </Content> 
   </ItemGroup>
+    <ItemGroup>
+      <Compile Remove="C:\Repos\azure-sdk-for-net\sdk\storage\Azure.Storage.Common\tests\Shared\TamperTransactionalHashPolicy.cs" />
```
What is this?  Also I think this is a local path.

jaschrep-msft replied:
Must have been generated by Visual Studio at some point. Removing it appears to do nothing.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/tests/TransactionalHashingTests.cs:
```diff
@@ -0,0 +1,653 @@
+            await using DisposingContainer test = await GetTestContainerAsync();
+
+            var blobName = GetNewBlobName();
+            await TransactionalHashingTestSkeletons.TestDownloadSuccessfulHashVerificationAsync(
```
Would it possible to simplify this test class?  I think the lambdas can be non-intuitive, and it could make these tests difficult to maintain.

jaschrep-msft replied:
The issue is that we need to test the exact same functionality across BlobClient, BlockBlobClient, AppendBlobClient, PageBlobClient, DataLakeFileClient, and ShareFileClient. The tests aren't the most intuitive things in the first place and cloning a dozen-ish of those tests 6 times is a maintainability nightmare. Genericizing the tests with lambdas to get the correct clients seemed to be the best approach. Perhaps there are optimizations to what the lambdas are but we need to inject resource creation and the operation to test.

seanmcc-msft replied:
At minimum, I think we need to add a lot of additional inline comments in the tests and shared test code to explain what is going on here.  These are some of the most complicated unit tests we have, and I'm concerned with maintainability if it isn't generally known how these work.

jaschrep-msft replied:
I will extensively docstring/inline the shared code as well as put some "why" descriptions on the methods/lambdas passed in.

kasobol-msft replied:
This is not a bad idea to create a template for hashing tests, but I'm not sure if current implementation reduces the codebase (i.e. I think that if we just multiplied the logic (copy-paste) the volume would be similar or even a bit smaller).

How about something like:
```
public abstract class TransactionalHashingTest {
     public TransactionalHashingTest (
      Delegate1 func1,
      Delegate2 func2,
      ...
     ) {
        this.func1 = func1;
        this.func2 = func2;
     }


     [Test]
     public async Task TestSomething() {
        await func1();
       
     }
     
     [Test]
     public async Task TestSomethingElse(){
       await func2();
     }
}

// and then have

public class BlockBlobHashingTests : TransactionalHashingTest {

   public BlockBlobHashingTests : base (func1, func2) {
   
   }

}

public class AppendBlobHashingTests : TransactionalHashingTest {

   public AppendBlobHashingTests : base (func1, func2) {
   
   }

}


public class ShareFileTransactionalHashingTest : TransactionalHashingTest {

   public ShareFileTransactionalHashingTest : base (func1, func2) {
   
   }

}

```

kasobol-msft replied:
@jaschrep-msft In this particular case , since we have an inheritance tree, these delegates might be abstract virtual methods perhaps.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/tests/TransactionalHashingTests.cs:
```diff
@@ -0,0 +1,653 @@
+        private Task<BlockBlobClient> MakeBlockBlobClient(BlobContainerClient container, BlobClientOptions testClientOptions)
+        {
+            container = InstrumentClient(new BlobContainerClient(container.Uri, GetNewSharedKeyCredentials(), testClientOptions));
+            return Task.FromResult(InstrumentClient(container.GetBlockBlobClient(GetNewBlobName())));
```
`container.GetBlockBlobClient()` isn't a async method, we could return a `BlockBlobClient` instead of a `Task<BlockBlobClient>` here

jaschrep-msft replied:
Block blobs are the only resource that don't need a `CreateAsync()` call to function. This method is the exception that just needs to fit the `Func` signature in tests.

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/UploadTransactionalHashingOptions.cs:
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Storage
```
Did the Arch Board decide on `Azure.Storage` instead of `Azure.Storage.Common`?

jaschrep-msft replied:
They did.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:
```diff
@@ -2516,9 +2622,211 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        [EditorBrowsable(EditorBrowsableState.Never)]
```
Is this supposed to be `EditorBrowsableState.Never`?

jaschrep-msft replied:
Nope. Good catch.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:
```diff
@@ -2516,9 +2622,211 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        [EditorBrowsable(EditorBrowsableState.Never)]
```
Same as above.

jaschrep-msft replied:
Nope. Good catch.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/Models/DataLakeFileAppendOptions.cs:
```diff
@@ -0,0 +1,30 @@
+namespace Azure.Storage.Files.DataLake.Models
+{
+    /// <summary>
+    /// Optional paratmers for appending data to a file.
```
Typo, and we could add a reference to `DataLakeFileClient.AppendAsync()` here.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/Models/DataLakeFileReadOptions.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Azure.Storage.Files.DataLake.Models
+{
+    /// <summary>
+    /// Optional paratmers for downloading some range of a file.
```
Typo, same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/tests/TransactionalHashingTests.cs:
```diff
@@ -0,0 +1,330 @@
+
+namespace Azure.Storage.Files.DataLake.Tests
+{
+    public class TransactionalHashingTests : DataLakeTestBase
```
Same as my comment on the Blob tests.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileDownloadOptions.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Azure.Storage.Files.Shares.Models
+{
+    /// <summary>
+    /// Optional paratmers for downloading some range of a file.
```
Typo.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileUploadOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        /// Constructs options instance using a stream as content source.
+        /// </summary>
+        /// <param name="stream">Content stream.</param>
+        public ShareFileUploadOptions(Stream stream)
```
Why is `stream` in the constructor instead of being settable?

jaschrep-msft replied:
I recall seeing a pattern in our libraries where required parameters are in options constructors instead of the method signature. e.g. the above instead of `Upload(Stream stream, ShareFileUploadOptions options)`. Which tends to be nice when introducing different required parameter sets; adding a `string path` alternative to providing a stream is just another constructor on the class instead of two new overloads on the client.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileUploadRangeOptions.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Azure.Storage.Files.Shares.Models
+{
+    /// <summary>
+    /// Optional paratmers for uploading a file range.
```
Typo.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4252,6 +4463,80 @@ private void SetNameFieldsIfNull()
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        [ForwardsClientCalls]
```
Same as my `DiagonsticScope` comment from before.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4252,6 +4463,80 @@ private void SetNameFieldsIfNull()
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        [ForwardsClientCalls]
```
Same as above.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/tests/TransactionalHashingTests.cs:
```diff
@@ -0,0 +1,304 @@
+
+namespace Azure.Storage.Files.Shares.Tests
+{
+    public class TransactionalHashingTests : FileTestBase
```
Same as my comment on the Blobs tests.

jaschrep-msft committed:
misc pr feedback

jaschrep-msft committed:
Fixed a docstring

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/DownloadTransactionalHashingOptions.cs:
```diff
@@ -0,0 +1,25 @@
+        /// </summary>
+        public TransactionalHashAlgorithm Algorithm { get; set; }
+
+        /// <summary>
+        /// Indicates that while the transactional hash should still be
+        /// requested on download, it should NOT be validated by this call.
+        /// Instead, the caller is responsible for validating the hash on the
+        /// returned <see cref="Response{T}"/>. If unsure, leave this to
+        /// default of false.
+        /// </summary>
+        public bool DeferValidation { get; set; }
```
Maybe we should have 
```
enum TransactionalHashingValidationMode {
  OnlyRetrieveHash,
  RetrieveAndValidateHash,
}
```

That boolean is confusing.

jaschrep-msft committed:
fixed diagnostic scope issue

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/Errors.cs:
```diff
@@ -46,6 +46,15 @@ public static ArgumentException InvalidService(char s)
             => new ArgumentException($"Cannot upload {streamLength} bytes with a maximum transfer size of {statedMaxBlockSize} bytes per block. Please increase the StorageTransferOptions.MaximumTransferSize to at least {necessaryMinBlockSize}.");
 
+        public static InvalidDataException HashMismatch(string hashHeaderName)
+            => new InvalidDataException($"{hashHeaderName} did not match recieved data.");
```
nit: "{hashHeaderName} did not match hash of recieved data."

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/Errors.cs:
```diff
@@ -46,6 +46,15 @@ public static ArgumentException InvalidService(char s)
+            => new InvalidDataException($"{hashHeaderName} did not match recieved data.");
+
+        public static ArgumentException CannotDeferTransactionalHashVerificationWithClientsideEncryption()
+            => new ArgumentException($"Cannot defer transactional hash verification. Hash goes to pre-decrypted contents, which are unavailable to caller.");
```
nit: "Cannot defer transactional hash verification. Hash goes to encrypted contents, which are unavailable to caller."

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/Errors.cs:
```diff
@@ -46,6 +46,15 @@ public static ArgumentException InvalidService(char s)
+            => new ArgumentException($"Cannot defer transactional hash verification. Hash goes to pre-decrypted contents, which are unavailable to caller.");
+
+        public static ArgumentException PrecalculatedHashNotSupportedOnSplit()
+            => new ArgumentException("Precalculated hash not supported when potentially partitioning an upload.");
```
when do we throw this?

jaschrep-msft replied:
Argument validation in PartitionedUploader constructor. Also start of ShareFileClient.UploadInternal, since it doesn't use the partitioned uploader for some reaosn.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/NonCryptographicHashAlgorithmHasher.cs:
```diff
@@ -0,0 +1,46 @@
+            if (stream is null)
+                throw new ArgumentNullException(nameof(stream));
+
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
```
nit. can we use constant?

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/StorageCrc64Calculator.cs:
```diff
@@ -0,0 +1,1052 @@
+    /// Port of https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Core/Util/Crc64.cs
+    /// and the code used by Azure Storage to compute a CRC-64 with custom polynomial.
+    /// </summary>
+    internal static class StorageCrc64Calculator
```
do we have extensive tests for this class?

jaschrep-msft replied:
I can search the v11 library for tests but this is a copy of a copy of the actual code in the service. I'm not entirely certain what would even be tested for something like this. We may need to reach out to FE if the v11 library just took their word for it.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/UploadTransactionalHashingOptions.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// Optional. For an upload API that ends up using a single REST call,
+        /// a precalculated hash for the upload contents can be provided to skip
+        /// recalculating. This hash will be ignored if the upload splits into
+        /// multiple REST calls and new ones will be calculated.
```
are we going to ignore or throw? if it gets ignored it's a trap.

jaschrep-msft replied:
We're supposed to throw. Likely this is an old doc that needs updating but I will check the code as well.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1091,43 +1108,146 @@ private void SetNameFieldsIfNull()
-            HttpRange requestedRange = range;
+            options ??= new BlobBaseDownloadOptions();
+
+            if ((options.TransactionalHashingOptions?.DeferValidation ?? false) && UsingClientSideEncryption)
+            {
+                throw Errors.CannotDeferTransactionalHashVerificationWithClientsideEncryption();
+            }
```
Clientside encryption will certainly detect a roundtrip bitflip, as the data will fail to correctly decrypt and be very obvious about that. But it won't detect it specifically on the upload or the download operation. Users need to know their upload was secure, especially when adding encryption blocks any previously possible attempt to analyze the bad data and see if it's reversible.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -854,6 +860,16 @@ public new BlockBlobClient WithEncryptionScope(string encryptionScope)
+                    // compute hash BEFORE attaching progress handler
+                    ContentHasher.GetHashResult hashResult = ContentHasher.GetHash(content, hashingOptions);
+                    // CRC not currently in generated code
+                    if (hashResult.StorageCrc64 != default)
+                    {
+                        throw new NotImplementedException("CRC64 support not implemented for PUT Blob.");
+                    }
```
Swagger currently does not support. I decided updating that could be put off for a separate PR.

jaschrep-msft committed:
Documented TransactionalHashingTestSkeletons.cs

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/BlobBaseDownloadOptions.cs:
```diff
@@ -0,0 +1,37 @@
+        /// </summary>
+        public DownloadTransactionalHashingOptions TransactionalHashingOptions { get; set; }
+
+        /// <summary>
+        /// Operation name for internal diagnostic scope use.
+        /// </summary>
+        internal string OperationName { get; set; }
```
pull out into XxxInternal signatures explicitly.

jaschrep-msft committed:
test comments

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1130,23 +1217,26 @@ public new AppendBlobClient WithEncryptionScope(string encryptionScope)
 
-                    content = content?.WithNoDispose().WithProgress(progressHandler);
+                    // compute hash BEFORE attaching progress handler
+                    ContentHasher.GetHashResult hashResult = ContentHasher.GetHash(content, options?.TransactionalHashingOptions);
```
Renamed to `GetHashOrDefault`. Documented to return `default` when the given hashing options are default, but still throw if the hash algorithm identifier on an options instance is invalid.

jaschrep-msft committed:
Some PR responses

Some options classes renamed
GetHashResult memory optimizations.
Disabled clientside crypto/hashing in combination

azure-sdk commented:
API changes have been detected in `Azure.Storage.Blobs`. You can review API changes [here](https://apiview.dev/Assemblies/Review/7e95569dda384bdbba740fcc5d3ee2cf?diffOnly=True&diffRevisionId=58b678a5e9c24cce9392c728db807e11)

azure-sdk commented:
API changes have been detected in `Azure.Storage.Common`. You can review API changes [here](https://apiview.dev/Assemblies/Review/68dcc61ccbf34a89a8030e05b7f4ab9b?diffOnly=True&diffRevisionId=d538d0158fe44f1d8ad45afdd4c0dfce)

azure-sdk commented:
API changes have been detected in `Azure.Storage.Files.DataLake`. You can review API changes [here](https://apiview.dev/Assemblies/Review/3b7db204dfa344e8a44fa002d9d90996?diffOnly=True&diffRevisionId=586ee6a5392a469d9f6337bd420adc0e)

jaschrep-msft committed:
new downloadto options have consistent API shape

azure-sdk commented:
API changes have been detected in `Azure.Storage.Blobs`. You can review API changes [here](https://apiview.dev/Assemblies/Review/7e95569dda384bdbba740fcc5d3ee2cf?diffOnly=True&diffRevisionId=a5c7d0ec98fd4ca4a8871e90f1015fab)

**API changes**
```diff
-         public BlobDownloadToOptions(string path);
-         public BlobDownloadToOptions(Stream targetStream);
+         public BlobDownloadToOptions();
-         public virtual Response DownloadTo(BlobDownloadToOptions options, CancellationToken cancellationToken = default);
+         public virtual Response DownloadTo(Stream destination, BlobDownloadToOptions options, CancellationToken cancellationToken = default);
+         public virtual Response DownloadTo(string path, BlobDownloadToOptions options, CancellationToken cancellationToken = default);
-         public virtual Task<Response> DownloadToAsync(BlobDownloadToOptions options, CancellationToken cancellationToken = default);
+         public virtual Task<Response> DownloadToAsync(Stream destination, BlobDownloadToOptions options, CancellationToken cancellationToken = default);
+         public virtual Task<Response> DownloadToAsync(string path, BlobDownloadToOptions options, CancellationToken cancellationToken = default);
```

azure-sdk commented:
API changes have been detected in `Azure.Storage.Files.DataLake`. You can review API changes [here](https://apiview.dev/Assemblies/Review/3b7db204dfa344e8a44fa002d9d90996?diffOnly=True&diffRevisionId=2d115cfa8c9846baa48d964ac87a2b1b)

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into feature/storage/transactional-crc

jaschrep-msft committed:
datalake hashing tests now use hns tenant

jaschrep-msft committed:
More PR comments

jaschrep-msft committed:
export api

jaschrep-msft committed:
new options bag APIs memory optimization

jaschrep-msft committed:
fixed failed tests

jaschrep-msft committed:
Initial hashing test conversion

Converted TransactionalHashingTestSkeletons (collection of static helper
methods) to TransactionalHashingTestBase : StorageTestBase. Tests are
fully defined on this class. Subclasses per-client will implement
abstract methods as instructions for uploads/downloads/etc., where
before they were lambdas passed to the static helper methods every
single time.

jaschrep-msft committed:
moved transactional hashing tests under new inheritance tree

jaschrep-msft committed:
deleted old hashing test session records

jaschrep-msft committed:
fixed playback bug and rerecorded to accomodate

jaschrep-msft committed:
Recorded inconclusive hashing tests

jaschrep-msft committed:
more recording inconclusive tests

jaschrep-msft committed:
quick fix

jaschrep-msft committed:
nitpicks/cleanup

jaschrep-msft committed:
ported v11 crc test

jaschrep-msft committed:
file header

jaschrep-msft committed:
reverted some auto IDE csproj changes

jaschrep-msft committed:
corrected a formating reversion

jaschrep-msft committed:
pr feedback

jaschrep-msft committed:
arg validation

jaschrep-msft committed:
pr feedback

kasobol-msft reviewed (approved):
Please create work item to further explore the CRC testing.

jaschrep-msft committed:
Changelog

jaschrep-msft committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into feature/storage/transactional-crc
